File: gr00t2/IsaacTeleop/src/core/schema_tests/cpp/test_full_body.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Unit tests for the generated FullBodyPosePico FlatBuffer message.

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>
#include <flatbuffers/flatbuffers.h>

// Include generated FlatBuffer headers.
#include <schema/full_body_generated.h>
#include <schema/timestamp_generated.h>

#include <memory>
#include <type_traits>

// =============================================================================
// Compile-time verification of FlatBuffer field IDs.
// These ensure schema field IDs remain stable across changes.
// VT values are computed as: (field_id + 2) * 2.
// =============================================================================
#define VT(field) (field + 2) * 2
static_assert(core::FullBodyPosePico::VT_JOINTS == VT(0));

static_assert(core::FullBodyPosePicoRecord::VT_DATA == VT(0));

// =============================================================================
// Compile-time verification of FlatBuffer field types.
// These ensure schema field types remain stable across changes.
// =============================================================================
#define TYPE(field) decltype(std::declval<core::FullBodyPosePico>().field())
static_assert(std::is_same_v<TYPE(joints), const core::BodyJointsPico*>);

// =============================================================================
// Compile-time verification of BodyJointPose struct.
// =============================================================================
static_assert(std::is_trivially_copyable_v<core::BodyJointPose>, "BodyJointPose should be a trivially copyable struct");

// =============================================================================
// Compile-time verification of BodyJointsPico struct.
// =============================================================================
static_assert(std::is_trivially_copyable_v<core::BodyJointsPico>, "BodyJointsPico should be a trivially copyable struct");

// BodyJointsPico should contain exactly 24 BodyJointPose entries.
static_assert(sizeof(core::BodyJointsPico) == 24 * sizeof(core::BodyJointPose),
              "BodyJointsPico should contain exactly 24 BodyJointPose entries");

// =============================================================================
// Compile-time verification of BodyJointPico enum.
// =============================================================================
static_assert(core::BodyJointPico_PELVIS == 0, "PELVIS should be index 0");
static_assert(core::BodyJointPico_RIGHT_HAND == 23, "RIGHT_HAND should be index 23");
static_assert(core::BodyJointPico_NUM_JOINTS == 24, "NUM_JOINTS should be 24");

// =============================================================================
// BodyJointPose Tests
// =============================================================================
TEST_CASE("BodyJointPose struct can be created and accessed", "[full_body][struct]")
{
    core::Point position(1.0f, 2.0f, 3.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose pose(position, orientation);
    core::BodyJointPose joint_pose(pose, true);

    SECTION("Pose values are accessible")
    {
        CHECK(joint_pose.pose().position().x() == 1.0f);
        CHECK(joint_pose.pose().position().y() == 2.0f);
        CHECK(joint_pose.pose().position().z() == 3.0f);
    }

    SECTION("is_valid is accessible")
    {
        CHECK(joint_pose.is_valid() == true);
    }
}

TEST_CASE("BodyJointPose default construction", "[full_body][struct]")
{
    core::BodyJointPose joint_pose;

    // Default values should be zero/false.
    CHECK(joint_pose.pose().position().x() == 0.0f);
    CHECK(joint_pose.pose().position().y() == 0.0f);
    CHECK(joint_pose.pose().position().z() == 0.0f);
    CHECK(joint_pose.is_valid() == false);
}

// =============================================================================
// BodyJointsPico Tests
// =============================================================================
TEST_CASE("BodyJointsPico struct has correct size", "[full_body][struct]")
{
    // BodyJointsPico should have exactly NUM_JOINTS entries (XR_BD_body_tracking).
    core::BodyJointsPico joints;
    CHECK(joints.joints()->size() == static_cast<size_t>(core::BodyJointPico_NUM_JOINTS));
}

TEST_CASE("BodyJointsPico can be accessed by index", "[full_body][struct]")
{
    core::BodyJointsPico joints;

    // Access first and last entries (returns pointers).
    const auto* first = (*joints.joints())[0];
    const auto* last = (*joints.joints())[core::BodyJointPico_NUM_JOINTS - 1];

    // Default values should be zero.
    CHECK(first->pose().position().x() == 0.0f);
    CHECK(last->pose().position().x() == 0.0f);
}

// =============================================================================
// FullBodyPosePicoT Tests
// =============================================================================
TEST_CASE("FullBodyPosePicoT default construction", "[full_body][native]")
{
    auto body_pose = std::make_unique<core::FullBodyPosePicoT>();

    // Default values.
    CHECK(body_pose->joints == nullptr);
}

TEST_CASE("FullBodyPosePicoT can store joints data", "[full_body][native]")
{
    auto body_pose = std::make_unique<core::FullBodyPosePicoT>();

    // Create and set joints.
    body_pose->joints = std::make_unique<core::BodyJointsPico>();

    CHECK(body_pose->joints->joints()->size() == static_cast<size_t>(core::BodyJointPico_NUM_JOINTS));
}

TEST_CASE("FullBodyPosePicoT joints can be mutated via flatbuffers Array", "[full_body][native]")
{
    auto body_pose = std::make_unique<core::FullBodyPosePicoT>();
    body_pose->joints = std::make_unique<core::BodyJointsPico>();

    // Create a joint pose.
    core::Point position(1.0f, 2.0f, 3.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose pose(position, orientation);
    core::BodyJointPose joint_pose(pose, true);

    // Mutate first joint
    body_pose->joints->mutable_joints()->Mutate(0, joint_pose);

    // Verify.
    const auto* first_joint = (*body_pose->joints->joints())[0];
    CHECK(first_joint->pose().position().x() == 1.0f);
    CHECK(first_joint->pose().position().y() == 2.0f);
    CHECK(first_joint->pose().position().z() == 3.0f);
    CHECK(first_joint->is_valid() == true);
}

TEST_CASE("FullBodyPosePicoT serialization and deserialization", "[full_body][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(4096);

    // Create FullBodyPosePicoT with all fields set.
    auto body_pose = std::make_unique<core::FullBodyPosePicoT>();
    body_pose->joints = std::make_unique<core::BodyJointsPico>();

    // Set a few joint poses
    core::Point position(1.5f, 2.5f, 3.5f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose pose(position, orientation);
    core::BodyJointPose joint_pose(pose, true);

    body_pose->joints->mutable_joints()->Mutate(0, joint_pose);

    // Serialize.
    auto offset = core::FullBodyPosePico::Pack(builder, body_pose.get());
    builder.Finish(offset);

    // Deserialize.
    auto buffer = builder.GetBufferPointer();
    auto deserialized = flatbuffers::GetRoot<core::FullBodyPosePico>(buffer);

    // Verify.
    CHECK(deserialized->joints()->joints()->size() == static_cast<size_t>(core::BodyJointPico_NUM_JOINTS));

    const auto* first_joint = (*deserialized->joints()->joints())[0];
    CHECK(first_joint->pose().position().x() == Catch::Approx(1.5f));
    CHECK(first_joint->pose().position().y() == Catch::Approx(2.5f));
    CHECK(first_joint->pose().position().z() == Catch::Approx(3.5f));
    CHECK(first_joint->is_valid() == true);
}

TEST_CASE("FullBodyPosePicoT can be unpacked from buffer", "[full_body][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(4096);

    // Create and serialize.
    auto original = std::make_unique<core::FullBodyPosePicoT>();
    original->joints = std::make_unique<core::BodyJointsPico>();

    // Set multiple joint poses (--gen-mutable provides mutable_joints()).
    for (size_t i = 0; i < 24; ++i)
    {
        core::Point position(static_cast<float>(i), static_cast<float>(i * 2), static_cast<float>(i * 3));
        core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
        core::Pose pose(position, orientation);
        core::BodyJointPose joint_pose(pose, true);
        original->joints->mutable_joints()->Mutate(i, joint_pose);
    }

    auto offset = core::FullBodyPosePico::Pack(builder, original.get());
    builder.Finish(offset);

    // Unpack to FullBodyPosePicoT.
    auto buffer = builder.GetBufferPointer();
    auto body_pose_fb = flatbuffers::GetRoot<core::FullBodyPosePico>(buffer);
    auto unpacked = std::make_unique<core::FullBodyPosePicoT>();
    body_pose_fb->UnPackTo(unpacked.get());

    // Check a few joints.
    const auto* joint_5 = (*unpacked->joints->joints())[5];
    CHECK(joint_5->pose().position().x() == Catch::Approx(5.0f));
    CHECK(joint_5->pose().position().y() == Catch::Approx(10.0f));
    CHECK(joint_5->pose().position().z() == Catch::Approx(15.0f));

    const auto* joint_23 = (*unpacked->joints->joints())[23];
    CHECK(joint_23->pose().position().x() == Catch::Approx(23.0f));
    CHECK(joint_23->pose().position().y() == Catch::Approx(46.0f));
    CHECK(joint_23->pose().position().z() == Catch::Approx(69.0f));
}

TEST_CASE("FullBodyPosePicoT all 24 joints can be set and verified", "[full_body][native]")
{
    auto body_pose = std::make_unique<core::FullBodyPosePicoT>();
    body_pose->joints = std::make_unique<core::BodyJointsPico>();

    // Set all 24 joints with unique positions (--gen-mutable provides mutable_joints()).
    for (size_t i = 0; i < 24; ++i)
    {
        core::Point position(static_cast<float>(i), 0.0f, 0.0f);
        core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
        core::Pose pose(position, orientation);
        core::BodyJointPose joint_pose(pose, true);
        body_pose->joints->mutable_joints()->Mutate(i, joint_pose);
    }

    // Verify all joints.
    for (size_t i = 0; i < 24; ++i)
    {
        const auto* joint = (*body_pose->joints->joints())[i];
        CHECK(joint->pose().position().x() == Catch::Approx(static_cast<float>(i)));
        CHECK(joint->is_valid() == true);
    }
}

// =============================================================================
// Body Joint Index Tests (XR_BD_body_tracking joint mapping)
// =============================================================================
TEST_CASE("FullBodyPosePicoT joint indices correspond to body parts", "[full_body][joints]")
{
    // This test documents the expected joint mapping from XR_BD_body_tracking.
    // Joint indices:
    //   0: Pelvis, 1-2: Left/Right Hip, 3: Spine1, 4-5: Left/Right Knee,
    //   6: Spine2, 7-8: Left/Right Ankle, 9: Spine3, 10-11: Left/Right Foot,
    //   12: Neck, 13-14: Left/Right Collar, 15: Head, 16-17: Left/Right Shoulder,
    //   18-19: Left/Right Elbow, 20-21: Left/Right Wrist, 22-23: Left/Right Hand

    auto body_pose = std::make_unique<core::FullBodyPosePicoT>();
    body_pose->joints = std::make_unique<core::BodyJointsPico>();

    // Verify we can access all expected joint indices.
    REQUIRE(body_pose->joints->joints()->size() == 24);

    // Access key joints by index.
    const auto* pelvis = (*body_pose->joints->joints())[0];
    const auto* head = (*body_pose->joints->joints())[15];
    const auto* left_hand = (*body_pose->joints->joints())[22];
    const auto* right_hand = (*body_pose->joints->joints())[23];

    CHECK(pelvis->pose().position().x() == 0.0f);
    CHECK(head->pose().position().x() == 0.0f);
    CHECK(left_hand->pose().position().x() == 0.0f);
    CHECK(right_hand->pose().position().x() == 0.0f);
}

// =============================================================================
// Edge Cases
// =============================================================================
TEST_CASE("FullBodyPosePicoT buffer size is reasonable", "[full_body][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    auto body_pose = std::make_unique<core::FullBodyPosePicoT>();
    body_pose->joints = std::make_unique<core::BodyJointsPico>();

    auto offset = core::FullBodyPosePico::Pack(builder, body_pose.get());
    builder.Finish(offset);

    // Buffer should be reasonably sized for 24 joints.
    // Each BodyJointPose is Pose (28 bytes) + bool (1 byte with padding) = ~32 bytes.
    // 24 joints * 32 bytes = ~768 bytes, plus overhead.
    CHECK(builder.GetSize() < 2000);
}

// =============================================================================
// BodyJointPico Enum Tests
// =============================================================================
TEST_CASE("BodyJointPico enum has correct values", "[full_body][enum]")
{
    // Verify all 24 enum values match the XR_BD_body_tracking joint indices.
    CHECK(core::BodyJointPico_PELVIS == 0);
    CHECK(core::BodyJointPico_LEFT_HIP == 1);
    CHECK(core::BodyJointPico_RIGHT_HIP == 2);
    CHECK(core::BodyJointPico_SPINE1 == 3);
    CHECK(core::BodyJointPico_LEFT_KNEE == 4);
    CHECK(core::BodyJointPico_RIGHT_KNEE == 5);
    CHECK(core::BodyJointPico_SPINE2 == 6);
    CHECK(core::BodyJointPico_LEFT_ANKLE == 7);
    CHECK(core::BodyJointPico_RIGHT_ANKLE == 8);
    CHECK(core::BodyJointPico_SPINE3 == 9);
    CHECK(core::BodyJointPico_LEFT_FOOT == 10);
    CHECK(core::BodyJointPico_RIGHT_FOOT == 11);
    CHECK(core::BodyJointPico_NECK == 12);
    CHECK(core::BodyJointPico_LEFT_COLLAR == 13);
    CHECK(core::BodyJointPico_RIGHT_COLLAR == 14);
    CHECK(core::BodyJointPico_HEAD == 15);
    CHECK(core::BodyJointPico_LEFT_SHOULDER == 16);
    CHECK(core::BodyJointPico_RIGHT_SHOULDER == 17);
    CHECK(core::BodyJointPico_LEFT_ELBOW == 18);
    CHECK(core::BodyJointPico_RIGHT_ELBOW == 19);
    CHECK(core::BodyJointPico_LEFT_WRIST == 20);
    CHECK(core::BodyJointPico_RIGHT_WRIST == 21);
    CHECK(core::BodyJointPico_LEFT_HAND == 22);
    CHECK(core::BodyJointPico_RIGHT_HAND == 23);
    CHECK(core::BodyJointPico_NUM_JOINTS == 24);
}

TEST_CASE("BodyJointPico enum can index BodyJointsPico", "[full_body][enum]")
{
    auto body_pose = std::make_unique<core::FullBodyPosePicoT>();
    body_pose->joints = std::make_unique<core::BodyJointsPico>();

    // Set specific joints using enum values (--gen-mutable provides mutable_joints()).
    // Set HEAD joint with a recognizable position.
    core::Point head_position(0.0f, 1.7f, 0.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose head_pose(head_position, orientation);
    core::BodyJointPose head_joint(head_pose, true);
    body_pose->joints->mutable_joints()->Mutate(core::BodyJointPico_HEAD, head_joint);

    // Set LEFT_HAND joint with a recognizable position.
    core::Point left_hand_position(-0.5f, 1.0f, 0.3f);
    core::Pose left_hand_pose(left_hand_position, orientation);
    core::BodyJointPose left_hand_joint(left_hand_pose, true);
    body_pose->joints->mutable_joints()->Mutate(core::BodyJointPico_LEFT_HAND, left_hand_joint);

    // Verify using enum values to access.
    const auto* head = (*body_pose->joints->joints())[core::BodyJointPico_HEAD];
    CHECK(head->pose().position().y() == Catch::Approx(1.7f));
    CHECK(head->is_valid() == true);

    const auto* left_hand = (*body_pose->joints->joints())[core::BodyJointPico_LEFT_HAND];
    CHECK(left_hand->pose().position().x() == Catch::Approx(-0.5f));
    CHECK(left_hand->is_valid() == true);
}

// =============================================================================
// FullBodyPosePicoRecord Tests (timestamp lives on the Record wrapper)
// =============================================================================
TEST_CASE("FullBodyPosePicoRecord serialization with DeviceDataTimestamp", "[full_body][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(4096);

    auto record = std::make_shared<core::FullBodyPosePicoRecordT>();
    record->data = std::make_shared<core::FullBodyPosePicoT>();
    record->data->joints = std::make_unique<core::BodyJointsPico>();
    record->timestamp = std::make_shared<core::DeviceDataTimestamp>(1000000000LL, 2000000000LL, 3000000000LL);

    auto offset = core::FullBodyPosePicoRecord::Pack(builder, record.get());
    builder.Finish(offset);

    auto deserialized = flatbuffers::GetRoot<core::FullBodyPosePicoRecord>(builder.GetBufferPointer());

    CHECK(deserialized->timestamp()->available_time_local_common_clock() == 1000000000LL);
    CHECK(deserialized->timestamp()->sample_time_local_common_clock() == 2000000000LL);
    CHECK(deserialized->timestamp()->sample_time_raw_device_clock() == 3000000000LL);
    CHECK(deserialized->data()->joints()->joints()->size() == static_cast<size_t>(core::BodyJointPico_NUM_JOINTS));
}

TEST_CASE("FullBodyPosePicoRecord can be unpacked with DeviceDataTimestamp", "[full_body][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(4096);

    auto original = std::make_shared<core::FullBodyPosePicoRecordT>();
    original->data = std::make_shared<core::FullBodyPosePicoT>();
    original->data->joints = std::make_unique<core::BodyJointsPico>();
    original->timestamp = std::make_shared<core::DeviceDataTimestamp>(111LL, 222LL, 333LL);

    auto offset = core::FullBodyPosePicoRecord::Pack(builder, original.get());
    builder.Finish(offset);

    auto fb = flatbuffers::GetRoot<core::FullBodyPosePicoRecord>(builder.GetBufferPointer());
    auto unpacked = std::make_shared<core::FullBodyPosePicoRecordT>();
    fb->UnPackTo(unpacked.get());

    CHECK(unpacked->timestamp->available_time_local_common_clock() == 111LL);
    CHECK(unpacked->timestamp->sample_time_local_common_clock() == 222LL);
    CHECK(unpacked->timestamp->sample_time_raw_device_clock() == 333LL);
    CHECK(unpacked->data->joints->joints()->size() == static_cast<size_t>(core::BodyJointPico_NUM_JOINTS));
}

File: gr00t2/IsaacTeleop/src/core/schema_tests/python/test_full_body.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Unit tests for FullBodyPosePicoT and related types in isaacteleop.schema.

FullBodyPosePicoT is a FlatBuffers table that represents full body pose data:
- joints: BodyJointsPico struct containing 24 BodyJointPose entries (XR_BD_body_tracking)

BodyJointsPico is a struct with a fixed-size array of 24 BodyJointPose entries.

BodyJointPose is a struct containing:
- pose: The Pose (position and orientation)
- is_valid: Whether this joint data is valid

Timestamps are carried by FullBodyPosePicoRecord, not FullBodyPosePicoT.

Joint indices follow XrBodyJointBD enum:
  0: Pelvis, 1-2: Left/Right Hip, 3: Spine1, 4-5: Left/Right Knee,
  6: Spine2, 7-8: Left/Right Ankle, 9: Spine3, 10-11: Left/Right Foot,
  12: Neck, 13-14: Left/Right Collar, 15: Head, 16-17: Left/Right Shoulder,
  18-19: Left/Right Elbow, 20-21: Left/Right Wrist, 22-23: Left/Right Hand
"""

import pytest

from isaacteleop.schema import (
    FullBodyPosePicoT,
    FullBodyPosePicoRecord,
    BodyJointsPico,
    BodyJointPose,
    BodyJointPico,
    Pose,
    Point,
    Quaternion,
    DeviceDataTimestamp,
)


class TestBodyJointPoseConstruction:
    """Tests for BodyJointPose construction."""

    def test_default_construction(self):
        """Test default construction creates BodyJointPose with default values."""
        joint_pose = BodyJointPose()

        assert joint_pose is not None
        # Default pose values should be zero.
        assert joint_pose.pose.position.x == 0.0
        assert joint_pose.pose.position.y == 0.0
        assert joint_pose.pose.position.z == 0.0
        assert joint_pose.is_valid is False

    def test_construction_with_values(self):
        """Test construction with position, orientation, and is_valid."""
        position = Point(1.0, 2.0, 3.0)
        orientation = Quaternion(0.0, 0.0, 0.0, 1.0)
        pose = Pose(position, orientation)
        joint_pose = BodyJointPose(pose, True)

        assert joint_pose.pose.position.x == pytest.approx(1.0)
        assert joint_pose.pose.position.y == pytest.approx(2.0)
        assert joint_pose.pose.position.z == pytest.approx(3.0)
        assert joint_pose.is_valid is True


class TestBodyJointPoseAccess:
    """Tests for BodyJointPose property access."""

    def test_pose_access(self):
        """Test accessing pose property."""
        position = Point(1.5, 2.5, 3.5)
        orientation = Quaternion(0.1, 0.2, 0.3, 0.9)
        pose = Pose(position, orientation)
        joint_pose = BodyJointPose(pose, True)

        assert joint_pose.pose.position.x == pytest.approx(1.5)
        assert joint_pose.pose.orientation.w == pytest.approx(0.9)

    def test_is_valid_access(self):
        """Test accessing is_valid property."""
        pose = Pose(Point(), Quaternion())
        joint_pose = BodyJointPose(pose, True)

        assert joint_pose.is_valid is True


class TestBodyJointPoseRepr:
    """Tests for BodyJointPose __repr__ method."""

    def test_repr(self):
        """Test __repr__ returns a meaningful string."""
        pose = Pose(Point(1.0, 2.0, 3.0), Quaternion(0.0, 0.0, 0.0, 1.0))
        joint_pose = BodyJointPose(pose, True)

        repr_str = repr(joint_pose)

        assert "BodyJointPose" in repr_str
        assert "Pose" in repr_str


class TestBodyJointsPicoStruct:
    """Tests for BodyJointsPico struct."""

    def test_joints_access(self):
        """Test accessing all 24 joints via joints() method."""
        body_joints = BodyJointsPico()

        for i in range(24):
            joint = body_joints.joints(i)
            assert joint is not None

    def test_joints_out_of_range(self):
        """Test that accessing out of range index raises IndexError."""
        body_joints = BodyJointsPico()

        with pytest.raises(IndexError):
            _ = body_joints.joints(24)


class TestBodyJointsPicoRepr:
    """Tests for BodyJointsPico __repr__ method."""

    def test_repr(self):
        """Test __repr__ returns a meaningful string."""
        body_joints = BodyJointsPico()

        repr_str = repr(body_joints)
        assert "BodyJointsPico" in repr_str


class TestFullBodyPosePicoTConstruction:
    """Tests for FullBodyPosePicoT construction and basic properties."""

    def test_default_construction(self):
        """Test default construction creates FullBodyPosePicoT with pre-populated joints."""
        body_pose = FullBodyPosePicoT()

        assert body_pose is not None
        assert body_pose.joints is not None

    def test_parameterized_construction(self):
        """Test construction with joints."""
        joints = BodyJointsPico()
        body_pose = FullBodyPosePicoT(joints)

        assert body_pose.joints is not None


class TestFullBodyPosePicoTRepr:
    """Tests for FullBodyPosePicoT __repr__ method."""

    def test_repr_default(self):
        """Test __repr__ with default construction."""
        body_pose = FullBodyPosePicoT()

        repr_str = repr(body_pose)
        assert "FullBodyPosePicoT" in repr_str


class TestBodyJointPicoEnum:
    """Tests for BodyJointPico enum (joint index mapping for XR_BD_body_tracking)."""

    def test_all_joint_values_exist(self):
        """Test that all expected BodyJointPico enum values exist."""
        assert BodyJointPico.PELVIS is not None
        assert BodyJointPico.LEFT_HIP is not None
        assert BodyJointPico.RIGHT_HIP is not None
        assert BodyJointPico.SPINE1 is not None
        assert BodyJointPico.LEFT_KNEE is not None
        assert BodyJointPico.RIGHT_KNEE is not None
        assert BodyJointPico.SPINE2 is not None
        assert BodyJointPico.LEFT_ANKLE is not None
        assert BodyJointPico.RIGHT_ANKLE is not None
        assert BodyJointPico.SPINE3 is not None
        assert BodyJointPico.LEFT_FOOT is not None
        assert BodyJointPico.RIGHT_FOOT is not None
        assert BodyJointPico.NECK is not None
        assert BodyJointPico.LEFT_COLLAR is not None
        assert BodyJointPico.RIGHT_COLLAR is not None
        assert BodyJointPico.HEAD is not None
        assert BodyJointPico.LEFT_SHOULDER is not None
        assert BodyJointPico.RIGHT_SHOULDER is not None
        assert BodyJointPico.LEFT_ELBOW is not None
        assert BodyJointPico.RIGHT_ELBOW is not None
        assert BodyJointPico.LEFT_WRIST is not None
        assert BodyJointPico.RIGHT_WRIST is not None
        assert BodyJointPico.LEFT_HAND is not None
        assert BodyJointPico.RIGHT_HAND is not None

    def test_joint_values(self):
        """Test that BodyJointPico values match expected indices."""
        assert int(BodyJointPico.PELVIS) == 0
        assert int(BodyJointPico.LEFT_HIP) == 1
        assert int(BodyJointPico.RIGHT_HIP) == 2
        assert int(BodyJointPico.SPINE1) == 3
        assert int(BodyJointPico.LEFT_KNEE) == 4
        assert int(BodyJointPico.RIGHT_KNEE) == 5
        assert int(BodyJointPico.SPINE2) == 6
        assert int(BodyJointPico.LEFT_ANKLE) == 7
        assert int(BodyJointPico.RIGHT_ANKLE) == 8
        assert int(BodyJointPico.SPINE3) == 9
        assert int(BodyJointPico.LEFT_FOOT) == 10
        assert int(BodyJointPico.RIGHT_FOOT) == 11
        assert int(BodyJointPico.NECK) == 12
        assert int(BodyJointPico.LEFT_COLLAR) == 13
        assert int(BodyJointPico.RIGHT_COLLAR) == 14
        assert int(BodyJointPico.HEAD) == 15
        assert int(BodyJointPico.LEFT_SHOULDER) == 16
        assert int(BodyJointPico.RIGHT_SHOULDER) == 17
        assert int(BodyJointPico.LEFT_ELBOW) == 18
        assert int(BodyJointPico.RIGHT_ELBOW) == 19
        assert int(BodyJointPico.LEFT_WRIST) == 20
        assert int(BodyJointPico.RIGHT_WRIST) == 21
        assert int(BodyJointPico.LEFT_HAND) == 22
        assert int(BodyJointPico.RIGHT_HAND) == 23

    def test_all_joint_indices_accessible(self):
        """Test that all BodyJointPico values can be used to index BodyJointsPico."""
        body_joints = BodyJointsPico()

        all_joints = [
            BodyJointPico.PELVIS,
            BodyJointPico.LEFT_HIP,
            BodyJointPico.RIGHT_HIP,
            BodyJointPico.SPINE1,
            BodyJointPico.LEFT_KNEE,
            BodyJointPico.RIGHT_KNEE,
            BodyJointPico.SPINE2,
            BodyJointPico.LEFT_ANKLE,
            BodyJointPico.RIGHT_ANKLE,
            BodyJointPico.SPINE3,
            BodyJointPico.LEFT_FOOT,
            BodyJointPico.RIGHT_FOOT,
            BodyJointPico.NECK,
            BodyJointPico.LEFT_COLLAR,
            BodyJointPico.RIGHT_COLLAR,
            BodyJointPico.HEAD,
            BodyJointPico.LEFT_SHOULDER,
            BodyJointPico.RIGHT_SHOULDER,
            BodyJointPico.LEFT_ELBOW,
            BodyJointPico.RIGHT_ELBOW,
            BodyJointPico.LEFT_WRIST,
            BodyJointPico.RIGHT_WRIST,
            BodyJointPico.LEFT_HAND,
            BodyJointPico.RIGHT_HAND,
        ]
        assert len(all_joints) == 24

        for joint in all_joints:
            joint_pose = body_joints.joints(int(joint))
            assert joint_pose is not None

    def test_enum_int_conversion(self):
        """Test that BodyJointPico values can be converted to integers."""
        assert int(BodyJointPico.PELVIS) == 0
        assert int(BodyJointPico.HEAD) == 15
        assert int(BodyJointPico.RIGHT_HAND) == 23


class TestFullBodyPosePicoRecordTimestamp:
    """Tests for FullBodyPosePicoRecord with DeviceDataTimestamp."""

    def test_construction_with_timestamp(self):
        """Test FullBodyPosePicoRecord carries DeviceDataTimestamp."""
        data = FullBodyPosePicoT()
        ts = DeviceDataTimestamp(1000000000, 2000000000, 3000000000)
        record = FullBodyPosePicoRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 1000000000
        assert record.timestamp.sample_time_local_common_clock == 2000000000
        assert record.timestamp.sample_time_raw_device_clock == 3000000000
        assert record.data is not None

    def test_default_construction(self):
        """Test default FullBodyPosePicoRecord has no data."""
        record = FullBodyPosePicoRecord()
        assert record.data is None

    def test_timestamp_fields(self):
        """Test all three DeviceDataTimestamp fields are accessible."""
        data = FullBodyPosePicoT()
        ts = DeviceDataTimestamp(111, 222, 333)
        record = FullBodyPosePicoRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 111
        assert record.timestamp.sample_time_local_common_clock == 222
        assert record.timestamp.sample_time_raw_device_clock == 333
